Refactor Program Management section of Admin Settings

Kafui123 opened the pull request:
## Issue
Issue: #797

## Changes
- Created a modal to update the information displayed for the programs on the homepage.
- Fields include program name, description, partner, contact email, contact name, and location.
- Implemented a add a new file feature, to change the image of the programs on the homepage.
- Saved changes button works as expected, and there are two ways to close the modal, cancel or cross button.
- Partner, contact email, contact name, location all stored in database.
- Program name, description, images and links  can be directly seen in the home screen.
- Added editable link for Instagram, Facebook and Berea URL.
- Refactored testing function

## Testing
- Checkout `refactorProgramMgmt`
- `./database/reset_database.sh test`
- `flask run`
- Go to Admin -> Settings
- Open up the Program Settings accordion
- Click on 3-4 programs and upload a png to 2 and a jpg to 2.
- Go to the landing page and make sure the landing page images for each program reflects what you uploaded.
- Go to `app/static/files/programattachments` and make sure that there is only one file for each program (1.jpg, 2.png, etc.)
- Once your testing has concluded delete all files that were just added and are not source controlled (pngs or any other different file types)
- Run `git restore .`
- If it does not properly add back all files that were initially there you may have to manually add them from GitHub or an earlier commit.

JohnCox2211 committed:
New Branch with changes from previous. Image
saving solution in route, currently get system
error.

JohnCox2211 committed:
changed coverImage to be a blobfield. cleaned up
some white space and useless comments and moved
some css to its proper location. Image solution
still doesn't work, will ask a higher power about
it on monday

JohnCox2211 committed:
More attempts for file handling, trying to modify
file handler, was trying to use userManagement.py

JohnCox2211 committed:
Deleted partner.py because it was not being used
except as a foreign key value in program to store
one charfield variable. Add some things to hopefully
handle files, not working. Updated the test data
to contain added data fields.

JohnCox2211 committed:
added some notes about what returns are for
some variables

Kafui123 committed:
modified the modal so we could store the images of the filepaths

Kafui123 committed:
modified image storing to database

Kafui123 committed:
we are on to something

Kafui123 committed:
Fixed Issuegit status !

Commit:
Picture appeared, at some point of time, but it disappeared again. Currently attempting to change the file name to the name of the program along with the file extension.

Kafui123 committed:
changes we have made so far

Commit:
Issue fixed, the modal works, the delete button does not close the modal and the msgToast is also in front of the modal and clearly visible.

Kafui123 committed:
updated code that reflects all the changes that work

AravDe committed:
Merge branch 'development' into refactorProgramMgmt

AravDe commented on app/logic/fileHandler.py:
```diff
@@ -2,18 +2,23 @@
+    def __init__(self, files=None, courseId=None, eventId=None, programId=None):
+        self.files = files 
         self.path = app.config['files']['base_path']
+        self.path1 = app.config['files']['image_path']    
```
Better Name

Kafui123 replied:
Fixed.

Kafui123 committed:
modified changes

Kafui123 committed:
updates
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

Commit:
Test suites almost fixed. Working on Mocking Filestorage object for testing.

Kafui123 committed:
updated code yet

Commit:
Added scroll feature to modal but now planning on removing again. Tests still not passed FileStorage issue.

Kafui123 committed:
modified changes

Kafui123 committed:
modified code
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

Commit:
Handing over to Kafui

Kafui123 committed:
modified changes

Kafui123 committed:
changes

eunsungw committed:
Changes to Kafui

Kafui123 committed:
modified changes

Kafui123 committed:
modified changes

Kafui123 committed:
updates for pr

Kafui123 committed:
Merge branch 'development' into refactorProgramMgmt

eunsungw committed:
Update for my local branch

Kafui123 committed:
final code ready for review

Kafui123 committed:
final code changes
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

Kafui123 committed:
Merge branch 'development' into refactorProgramMgmt

BrianRamsay commented on app/config/default.yml:
```diff
@@ -25,9 +25,12 @@ ldap:
   email_attachment_path: 'emailattachments'
   event_attachment_path: 'eventattachments'
   course_attachment_path: 'courseattachments'
+  program_image_path: 'programimages'
```
Looks like we don't use this one

eunsungw replied:
We removed the code.

BrianRamsay commented on app/controllers/admin/userManagement.py:
```diff
@@ -62,29 +65,37 @@ def addProgramManagers():
     except Exception as e:
         print(e)
         flash('Error while removing a manager.','warning')
         abort(500,"Error while trying to remove a manager.")
```
These functions aren't used anymore. Let's get rid of them

eunsungw replied:
Removed the function.

BrianRamsay commented on app/controllers/admin/userManagement.py:
```diff
@@ -62,29 +65,37 @@ def addProgramManagers():
+@admin_bp.route('/deleteProgramFile', methods=['POST'])
+def deleteProgramFile():
+    fileData= request.form
+    programFile=FileHandler(programId=fileData["databaseId"])
```
`databaseId` should be `programId`, because it is the id of the program.

eunsungw replied:
Changed the ID.

BrianRamsay commented on app/controllers/admin/userManagement.py:
```diff
@@ -62,29 +65,37 @@ def addProgramManagers():
 
+@admin_bp.route('/deleteProgramFile', methods=['POST'])
+def deleteProgramFile():
+    fileData= request.form
```
I wouldn't use a variable here. The other two lines are getting data out of the form, so it's fine to use `request.form`

eunsungw replied:
We got rid of the variable and used request.form

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -2,18 +2,23 @@
             self.path = os.path.join(self.path, app.config['files']['event_attachment_path'])
-    
+        elif programId:
+            self.image_path = os.path.join(self.image_path, app.config['files']['landing_page_path']) 
```
Let's use `self.path` like the other cases

Kafui123 replied:
We made the changes.

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -25,17 +30,25 @@ def makeDirectory(self):
+            if self.eventId or self.courseId:
+                filePath = (os.path.join(self.path, newfilename))
+            elif self.programId:
+                filePath = (os.path.join(self.image_path, newfilename))
```
This can go back to the old code, once you are using `self.path`

Kafui123 replied:
Done.

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -25,17 +30,25 @@ def makeDirectory(self):
+      
+        try: 
+            if not isinstance(self.files, list):
+                self.files = [self.files]          
```
Handling the input to the class and translating that to member variables is what the `__init__` method is for. Do this there.

Kafui123 replied:
Fixed

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -49,11 +62,28 @@ def saveFiles(self, saveOriginalFile=None):
+                    if not  isFileInProgram:
+                        AttachmentUpload.create(program=self.programId, fileName=file.filename)                      
+                        name = Program.get(Program.id == self.programId)
+                        file_type = file.filename.split('.')[1] 
```
We want to get the LAST bit after a `.`, not the second piece, in case they name something like "habitat.humanity.jpg"

Kafui123 replied:
Resolved.

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -49,11 +62,28 @@ def saveFiles(self, saveOriginalFile=None):
+                        name = Program.get(Program.id == self.programId)
+                        file_type = file.filename.split('.')[1] 
+                        current_programName = f"{str(name.programName)}.{file_type}"
+                        pattern = '*' + str(name.programName) + '*'
```
`name.programName` is always a string (database field is a varchar), so the cast isn't necessary.

Kafui123 replied:
Fixed.

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -49,11 +62,28 @@ def saveFiles(self, saveOriginalFile=None):
+                        full_pattern = os.path.join(self.image_path, pattern)
+                        files_to_delete = glob.glob(full_pattern)
+                       
+
```
Remove extra newline

Kafui123 replied:
We got rid of the extra new line

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -49,11 +62,28 @@ def saveFiles(self, saveOriginalFile=None):
                         AttachmentUpload.create(course=self.courseId, fileName=file.filename)
                         saveFileToFilesystem = file.filename
+                elif self.programId:
+                    isFileInProgram = AttachmentUpload.select().where(AttachmentUpload.program == self.programId, AttachmentUpload.fileName == file.filename).exists()
```
This check means you won't save different files with the same filename. Managing the database should mirror managing the filesystem, where you remove everything for the program and then save the new one.

Kafui123 replied:
Fixed.

BrianRamsay commented on app/logic/userManagement.py:
```diff
@@ -38,20 +39,45 @@ def removeCeltsStudentStaff(user):
-
-def changeProgramInfo(newProgramName, newContactEmail, newContactName, newLocation, programId):
-    """Updates the program info with a new sender and email."""
+def changeProgramInfo(programId, attachment, programName= None, programDescription = None, partner = None, contactEmail=None, contactName= None, location = None,instagramUrl = None, facebookUrl = None, bereaUrl = None): 
```
Let's format this so that the arguments are aligned

Kafui123 replied:
We have aligned the arguments.

BrianRamsay commented on app/models/program.py:
```diff
@@ -1,21 +1,21 @@
     contactName = CharField(null=True,default='')
     contactEmail = CharField(null=True,default='')
     defaultLocation = CharField(null=True,default='')
+    coverImage = BlobField(null=True)
```
If we don't use the field we should get rid of it.

Kafui123 replied:
CoverImage is being used as it is a foreign key of program in the database. Therefore, coverImage is used to store the attachment that contains the 'modalProgramImage'.

BrianRamsay commented on app/static/css/programManagement.css:
```diff
@@ -0,0 +1,57 @@
+    width: 10px;
+    height: 10px;
+  }
+}
```
If this is necessary we should be trying to use bootstrap's grid system, otherwise let's get rid of it.

Kafui123 replied:
we got rid of it because it was not necessary

BrianRamsay commented on app/static/js/base.js:
```diff
@@ -191,6 +191,7 @@ function handleFileSelection(fileInputId){
         $(fullTrashId).data("file", file);
         $(fullTrashId).data("file-container-id", attachedObjectContainerId);
         $(fullTrashId).on("click", function() {
+          console.log("hehe")
```
no debug output

Kafui123 replied:
Got rid of the debug output.

BrianRamsay commented on app/static/js/displayFilesMacro.js:
```diff
@@ -1,27 +1,28 @@
+    //         }
+    //     });
+    // });
 
```
We still need this code in other places.

eunsungw replied:
Uncommented the code.

BrianRamsay commented on app/static/js/userManagement.js:
```diff
@@ -63,6 +63,20 @@ function submitRequest(method, username){
+  document.getElementById('location').value = location;
+  document.getElementById('instagramUrl').value = instagramUrl;
+  document.getElementById('facebookUrl').value = facebookUrl;
+  document.getElementById('bereaUrl').value = bereaUrl;
```
We should use jquery like we do everywhere else

BrianRamsay replied:
`$("#programName").val(programName)`

Kafui123 replied:
We used jQuery.

BrianRamsay commented on app/templates/admin/userManagement.html:
```diff
@@ -114,36 +121,124 @@ <h3 class="accordion-header" id="headingTwo">
+                      type="module"> import { populateModal } from "/static/js/userManagement.js";
+                      window.programModal = function(programName, programDescription, partner, contactEmail, contactName, location, programId, instagramUrl, facebookUrl, bereaUrl){
+                      populateModal(programName, programDescription, partner, contactEmail, contactName, location, programId, instagramUrl, facebookUrl, bereaUrl)
+                      }
```
If we store the program info as a data element `data-programInfo="{{ programData }}"`, then we can add an onclick to all the buttons from javascript that grabs this data from whichever button is the `event.target`.  The snag is getting a JSON representation of the dictionary for the Program model object. I'm not sure how to do that. `model_to_dict` or some other method.

eunsungw replied:
We utilized JSON representation from the JS to directly populate the modal fields.

BrianRamsay commented on app/templates/admin/userManagement.html:
```diff
@@ -114,36 +121,124 @@ <h3 class="accordion-header" id="headingTwo">
+        <div class="modal-body">
+            <tr>
+              <!-- Left modal column -->
+              <td class="cellSpacing">
```
`<td>` is something that tables use. You want divs (probably), with bootstrap rows and cols

Kafui123 replied:
Changed it to divs

BrianRamsay commented on app/templates/admin/userManagement.html:
```diff
@@ -114,36 +121,124 @@ <h3 class="accordion-header" id="headingTwo">
+                </div>
+                <div class="mb-3">
+                  <label for="location" class="form-label">Location</label>
+                  <input type="text" class="form-control" id="location" name="location">
```
Let's rename to Default Event Location

Kafui123 replied:
Changed to Default Event Location

BrianRamsay reviewed (changes requested):
The UI needs a little work. Buttons may be better as "Edit Details", Program names shouldn't be `<input>` tags, which should fix the table column width. Buttons should be normal sized.

esw0624 commented:
#797

eunsungw committed:
Update

Kafui123 committed:
userpr

Kafui123 committed:
modified changes

Kafui123 committed:
checking to see if my commits are being sent to the remote bramch

eunsungw committed:
Update for branch

Kafui123 committed:
this commit reflects all updated changes that my partner, eun can pull from

Kafui123 committed:
necessary merge to get in sync with my remotebranch
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

Kafui123 committed:
final code changes for July 25

Kafui123 committed:
commit for friday july -26 (most recent code update)

eunsungw committed:
Merge branch 'development' into refactorProgramMgmt

Kafui123 committed:
most recent code on july 29th

Kafui123 committed:
most recent code for july 30th

Kafui123 committed:
most recent code change for july 30th
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

Kafui123 committed:
final code ready for pr review july 31st

BrianRamsay committed:
Merge branch 'development' into refactorProgramMgmt

BrianRamsay commented on app/controllers/admin/userManagement.py:
```diff
@@ -52,39 +54,38 @@ def manageUsers():
     except Exception as e:
         print(e)
         flash('Error while trying to add a manager.','warning')
         abort(500,"'Error while trying to add a manager.'")
```
This one also is not used (I'm assuming, because it calls itself, which would break).

Kafui123 replied:
Resolved.

BrianRamsay commented on app/logic/fileHandler.py:
```diff
@@ -25,17 +32,21 @@ def makeDirectory(self):
+            if self.eventId or self.courseId:
+                filePath = (os.path.join(self.path, newfilename))
+            elif self.programId:
+                filePath = (os.path.join(self.path, newfilename))
```
We don't need any conditional, because the code is the same in both cases.

Kafui123 replied:
Fixed.

BrianRamsay commented on app/logic/userManagement.py:
```diff
@@ -38,20 +39,53 @@ def removeCeltsStudentStaff(user):
+        Program.coverImage: attachment,
+        Program.instagramUrl:instagramUrl,
+        Program.facebookUrl: facebookUrl,
+    Program.bereaUrl: bereaUrl
```
These indents should all be at the same level

Kafui123 replied:
Fixed.

BrianRamsay reviewed (dismissed):
We forgot about the case when a program name changes. The current logic means it can no longer find the uploaded image.

Let's store the file path in coverImage (like we do in AttachmentUpload). This way, when the name changes we don't have to do anything. This will mean changing how the image is referenced on the home page, as well as how we delete images when we upload a new one.

The table for the programs is a little narrow. 350px minimum width would fit better. Also, the header Program should be centered.

Kafui123 committed:
final code changes for July 31st

Kafui123 committed:
final code updates for July 31st

Kafui123 committed:
merge necessary to update my local branch to my remote branch
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

Kafui123 committed:
most recent code update for august 2nd

bledsoef committed:
Merge branch 'development' into refactorProgramMgmt

WackyWeaver commented:
For the issue with the cover images of programs, it seems like uploaded photos are not stored, but their information is saved to the file under the program's name on the landing page. The field "coverImage" under Program still stores the original file's name. 

When the landing page loads, it makes the assumption that each program has the image equal to it's program. I need to keep the photo somewhere different, or I need to correct the assumption that landingPage.py is making.

WackyWeaver committed:
Switch the landingPage image names to program IDs

WackyWeaver committed:
Merge branch 'refactorProgramMgmt' of https://github.com/BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

WackyWeaver commented:
Since I am not the owner of this PR, I could not directly address the comment on the images from Brian. I found a solution to it! By using the program ID instead of the program name to rename the landing page images, renaming the programs no longer causes issue. The program cover images need to renamed to their respective IDs to display. 

Also, the UI for the program management still needs to be improved.

bledsoef committed:
Merge branch 'development' into refactorProgramMgmt

bledsoef committed:
fixed crappy merge conflicts that were preventing the database from resetting

WackyWeaver committed:
fixed the text blur and size of program settings modal

WackyWeaver committed:
added all program photos with id

WackyWeaver committed:
fixed unit landingPage unit test

WackyWeaver committed:
removed coverImage field and fixed unit tests

vungc reviewed (dismissed):
When uploading one of the photos, it does not actually update on the home page but it still give the success message (below is the example image that did not upload)
![image](https://github.com/user-attachments/assets/b20d0f81-3868-4cf9-bdc5-47a87998329c)
Also, I am able to select multiple images to upload, but only the first one will be the one updated on the homepage. It might be a good idea to disable the upload button or something like that so users are not confused.
![image](https://github.com/user-attachments/assets/62da023a-583e-4953-bb7c-71ac9c3e510a)

WackyWeaver committed:
test_image addedq

WackyWeaver committed:
test_image added

bledsoef committed:
created a sql script and updated file handler logic

bledsoef committed:
fixed sql script

bledsoef committed:
updated image directory and got landing page properly displaying

bledsoef committed:
moved the images to a new location

bledsoef committed:
changed upload image verbage

WackyWeaver committed:
replaces selected image in program modal

bledsoef committed:
worked on file handler for program attachments

bledsoef committed:
removed some suspect code

WackyWeaver committed:
working on test case

WackyWeaver committed:
Merge branch 'refactorProgramMgmt' of github.com:BCStudentSoftwareDevTeam/celts into refactorProgramMgmt

WackyWeaver committed:
removed print statement in testing suite

WackyWeaver committed:
added program photos to test data

WackyWeaver committed:
fixed typo

bledsoef committed:
created new testing ssuite

bledsoef committed:
modified test data to handle inserting program files

bledsoef committed:
refactored update program tests

WackyWeaver committed:
finishing touches on unit tests

thealphagurlux commented:
When trying to upload a picture for a event, if you exit out of the modal and open it the file name will still be there and the same file will still be uploaded if saved.

I suggest clearing the picture and file name when the user exits the modal.

<img width="608" alt="image" src="https://github.com/user-attachments/assets/618f602b-bb88-4698-8798-056ecc09eae8">

thealphagurlux commented:
For the following programs, when trying to change the event picture in settings the picture doesn't change:

-Berea Teen Mentoring
-Hispanic Outreach Program
-People Who Care

WackyWeaver committed:
merged the doc ready functionality

bledsoef committed:
fixed merge conflicts

bledsoef committed:
removed file testing

bledsoef commented:
> For the following programs, when trying to change the event picture in settings the picture doesn't change:
> 
> -Berea Teen Mentoring -Hispanic Outreach Program -People Who Care

@thealphagurlux I double checked these on my machine and Anna's and we could not replicate the issue you found.

bledsoef committed:
clear file contents on modal open

thealphagurlux commented:
> > For the following programs, when trying to change the event picture in settings the picture doesn't change:
> > -Berea Teen Mentoring -Hispanic Outreach Program -People Who Care
> 
> @thealphagurlux I double checked these on my machine and Anna's and we could not replicate the issue you found.

thealphagurlux commented:
> > > For the following programs, when trying to change the event picture in settings the picture doesn't change:
> > > -Berea Teen Mentoring -Hispanic Outreach Program -People Who Care
> > 
> > 
> > @thealphagurlux I double checked these on my machine and Anna's and we could not replicate the issue you found.

I had not reset my database or restored the git branch while looking for bugs. This is what was causing the issue.

WackyWeaver committed:
ignored unwanted toast msg

Mbeweg commented on app/controllers/admin/userManagement.py:
```diff
@@ -48,43 +50,32 @@ def manageUsers():
+            filePaths = FileHandler(programId=programID).retrievePath(associatedAttachments) 
+           
+          
+            file_paths = {filename: path_info[0] for filename, path_info in filePaths.items()} 
```
The `"filePaths" `variable and the `"file_paths" `variable should be both camelcase. The also should have different variable names.

WackyWeaver replied:
Honestly both lines seem to be assigning variables to objects or dictionaries that are not used in the rest of the function. So I decided to scrap them.

Karina-Agliullova commented on app/logic/fileHandler.py:
```diff
@@ -49,11 +58,27 @@ def saveFiles(self, saveOriginalFile=None):
+                    fileType = file.filename.split('.')[-1]
+                    fileName = f"{self.programId}.{fileType}"
+                    AttachmentUpload.create(program=self.programId, fileName=fileName)
+                    current_programID = fileName
```
should be camelCase

WackyWeaver replied:
Done!

Mbeweg commented on app/controllers/admin/userManagement.py:
```diff
@@ -48,43 +50,32 @@ def manageUsers():
         except Exception as e:
-            print(e)
-            flash('Error while updating program info.','warning')
+            print("error: ", e)
```
Get rid of the ``"print("error: ", e)" `we don't need it.

WackyWeaver replied:
Removed

Karina-Agliullova commented on app/static/js/base.js:
```diff
@@ -212,7 +218,12 @@ function handleFileSelection(fileInputId){
+          $(objectContainerId).html(fileHTML)
+        }
+        else{
         msgToast("File with filename '" + file.name + "' has already been added to this event")
```
indentation

WackyWeaver replied:
Added!

Mbeweg commented on app/logic/fileHandler.py:
```diff
@@ -49,11 +58,27 @@ def saveFiles(self, saveOriginalFile=None):
+                    AttachmentUpload.create(program=self.programId, fileName=fileName)
+                    current_programID = fileName
+                    saveFileToFilesystem = current_programID
+
```
I think using snake_case for variable naming is better. 
`deleteFileObject `to `delete_file_object.`
`fileType` to `file_type.`
`fileName` to `file_name.`

WackyWeaver replied:
I remember we discussed this in-person last week. The code base often uses camelCase instead of snake case.

WackyWeaver committed:
made all of the requested changes from last review

Karina-Agliullova reviewed (approved):
Looking good!

WackyWeaver committed:
Merge branch 'development' into refactorProgramMgmt
